Migrate the configuration to .yaml files

slaufmann opened the pull request:
The main goal of this PR is to migrate the configuration from python imports to .yaml files. This includes the custom configuration (search parameters, mail setup, etc.) as well as the site configurations.

In addition to just achieving this awesome goal some new sites were added, some old were fixed and reactivated.

And because it's a new year I topped that of with a new configuration feature that let's you choose which sites should not be crawled.
If that does not convince you to merge this PR I don't know what will. :)

slaufmann committed:
Use YAML as config file format (mail setup so far)

To achieve a better seperation of code and data the config settings are migrated to
YAML files. So far only the mail settings are read from config.yaml.

slaufmann committed:
Wrap search criteria into seperate class "SearchCOnfig".

In order to migrate the configuration to a .yaml file the search criteria are now
wrapped in the class "SearchConfig". This allows them to be passed along as a single
object.

slaufmann committed:
Read districts to search for at Gewobag and DeGeWo from .yaml config file

The district lists to search for at sites of Gewobag and DeGeWo are also read from
the .yaml config file. Selection of a subset of districts can be done via
commenting/uncommenting.

slaufmann committed:
Migrate site configuration to .yaml file

In order to migrate all configuration to .yaml files this is a first step to read
website configurations from the "sites.yaml" file. Parsing of websites and expose URLs
works but extraction of expose details is not yet implemented.

slaufmann committed:
Extend available site configurations, add format arguments for site URLs

All of the sites that were available in sites.py moved to sites.yaml. To preserve
functionality a URL can contain format arguments that are replaced with the values
as given in the search section of config.yaml.
Still the parsing of expose details is not implemented.

slaufmann committed:
Fix success-str, none-str and expose-url-pattern for DeGeWo in sites.yaml

slaufmann committed:
Move debug print to code path controlled by the --debug option

slaufmann committed:
Fix success-str for Wohnungsbaugenossenschaft Zentrum e.G. in sites.yaml

slaufmann committed:
Add expose-details to Wohnungsbaugenossenschaft DPF e.G., Gewobag and DeGeWo

slaufmann committed:
Extend configuration by a blocklist for sites that will not be crawled.

The configuration file config.yaml now features a section "site-blocklist" where
sites can be listed that should NOT be crawled. For these sites a simple log
message is shown instead.
The reason to implement a blocklist instead of a allowlist is so that newly added sites
are crawled by default.

slaufmann committed:
Remove sites.py as well as config-default.py since their content was migrated to yaml

The custom configuration as well as the site configurations were fully migrated to yaml
files. Therefore the old config files are no longer needed.

grandchild commented on config-default.yaml:
```diff
@@ -0,0 +1,78 @@
+
+# Configure your flat criteria. Not all fields are used by all sites.
+search:
+    rooms_min: 4
```
YAML uses 2 spaces as indentation. The YAML indentation convention overrides the python indentation rule for me.

grandchild commented on crawler.py:
```diff
@@ -32,16 +32,16 @@
 import sendmail
-from sites import sites as site_configs
-from config import MailConfig
+import yaml
```
This should be reflected in requirements.txt -- I guess `pyyaml` is what you want? (There is also `oyaml` if ordered maps is what you need.)

grandchild commented on config-default.yaml:
```diff
@@ -0,0 +1,78 @@
+        - "treptow-koepenick"
+        - "treptow-koepenick-adlershof"
+        - "treptow-koepenick-alt-treptow"
+---
```
You're separating this into documents, but there is no initial document.

I'd remove these altogether. I'd use document separators if I _had_ to save several _similar_ documents in the same file for whatever reason. The parts here are just sections, and that's not really what (in my experience) yaml document separators `---` are used for.

grandchild replied:
If you really want to visually separate them, use more empty lines or `#######` or `# ---------` or something like this.

slaufmann replied:
I was merely the fact that I thought it had to be separated this way. In my opinion the indentation is good enough for the visual separation.

grandchild commented on crawler.py:
```diff
@@ -81,21 +82,64 @@
+
+    def __init__(self):
+        with open("config.yaml", "r") as config_file:
+            for data in yaml.load_all(config_file):
```
I think you can use `yaml.safe_load` here. The default loader can load Python objects that execute code, that's not needed here.

slaufmann replied:
Yap, you're right.

grandchild commented on config-default.yaml:
```diff
@@ -0,0 +1,78 @@
+        - "64" # Reinickendorf
+        - "71" # Pankow
+# Wohnberechtigungsschein (If you don't know what this is leave it No)
+    wbs: No
```
Why is this a string and not an actual boolean? It'd also simplify the loading code.

Also I'd place it up with the other simple configuration items, not between these two huge lists.

grandchild replied:
Oh, wait, this is an actual boolean. But YAML 1.2 removes these other boolean representations, and only allows "true" and "false". Some libraries chose to still accept "yes/no" and (I think) "on/off" as well, but better to stay compatible and use "true/false"

grandchild commented on sendmail.py:
```diff
@@ -46,7 +46,23 @@
+class MailConfig:
+    def __init__(self):
+        with open("config.yaml", "r") as config_file:
+            for data in yaml.load_all(config_file):
```
Definitely `yaml.safe_load` here.

slaufmann replied:
Right again, changed it here too.

grandchild commented on crawler.py:
```diff
@@ -81,21 +82,64 @@
+                        yaml_data["gewobag_districts"]
+                    )
+                if "site-blocklist" in data:
+                    config["site-blocklist"] = data["site-blocklist"]
```
I'm wondering why you're loading these separately. I think it's because you want to parse the district lists with the separators, but it's too high a price.

I'd load the config all into one dict, with basically a oneliner, and then create an additional config key with the joined lists.

Roughly:
```python
config = yaml.safe_load(yaml_config)
config["degewo_districts_urljoined"] = "%2C+".join(config["degewo_districts"])
```

slaufmann replied:
I just learned that this needs the copy() constructor so that the dictionary contents are actually copied but now it works fine and looks much better in the code.

grandchild commented on crawler.py:
```diff
@@ -81,21 +82,64 @@
+                    )
+                if "site-blocklist" in data:
+                    config["site-blocklist"] = data["site-blocklist"]
+            config_file.close()
```
No need to `close()` the file, you're using a `with open()` for that. When the `with` part exits, it does all cleanup that `open()` needs.

grandchild commented on crawler.py:
```diff
@@ -81,21 +82,64 @@
     """
 
-    def __init__(self, config={}):
+    def __init__(self, config={}, search_config=None):
```
If it gains a `search_config` argument
  a) the `config` argument should be renamed to `site_config`
  b) the new argument should be mentioned in the docstring above.

slaufmann replied:
Good point, the "config" member of the "Site" class should stay unchanged though, since it implies being the "site_config". What do you think?

grandchild reviewed (changes requested):
Wow! Thanks 👍🏾 

I definitely like the yaml config, it's a bit cleaner.

There are a few issues still, and I think having two configs objects in the code is a bit overkill. I'd rather have one config class with site-config stuff embedded. (Strictly speaking about `crawler.py` -- I think the config files should stay split up.)

There's more, but I have to "go" to work ;) Will get back to this later.

slaufmann committed:
Change indentation of .yaml files to common standard of 2 spaces per indentation level

slaufmann committed:
Add pyyaml to list of requirements. Version 3.13 was chosen because it is the one I use.

slaufmann committed:
Different sections of the config file do not need to be separated by "---".

slaufmann committed:
Use true/false for boolean entry "wbs"

slaufmann committed:
No need to use close() on a file opened using the "with ... as" statement

slaufmann committed:
Replace file name in docstring

slaufmann committed:
Rename argument for clarification, update docstring

slaufmann committed:
Use copy constructor to load config data from file

Instead of populating all fields of the config dict in the SearchConfig class a copy
constructor is used. Some config entries which are derived by data from the file are
still created manually.

slaufmann committed:
Use defaultdict initilization for search_config dict to avoid key errors.

slaufmann committed:
No specific url_args config entry needed

It turns out that the url_args config entry is not needed. Every given URL can be
combined with the search_config dict and will not be altered if none of the dict keys
is found in the URL.

slaufmann committed:
Use safe_load_all function to prevent the Loader from executing Python code.

The safe_load* functions from the yaml package prevent the execution of Python code by
the loader which is preferred over the default loader when this is not needed.

slaufmann committed:
Fix error introduced by manual merge, name of expose_url_pattern field

slaufmann committed:
Use safe_load_all function in sendmail.py too

slaufmann commented:
I worked on the PR to honor your review comments. I also rebased on the current state of main so it should be ready for merging now.

grandchild commented on sendmail.py:
```diff
@@ -52,7 +52,7 @@
         with open("config.yaml", "r") as config_file:
-            for data in yaml.load_all(config_file):
+            for data in yaml.safe_load_all(config_file):
                 if "mail" in data:
```
does this actually work? `safe_load_all` should return a list, so the `"mail" in data` should be `False`, no?

grandchild replied:
i'd use plain `safe_load()`.

grandchild replied:
oh, sorry, it's a `for` loop... sure.

grandchild replied:
but yeah, you can lose the `for` loop with `safe_load()`.

grandchild commented on crawler.py:
```diff
@@ -84,21 +85,55 @@
+class SearchConfig:
+    def __init__(self):
+        with open("config.yaml", "r") as config_file:
+            for data in yaml.safe_load_all(config_file):
+                if "search" in data:
```
Same here, the `for` loop can be removed, since there's now only one document left.

slaufmann committed:
Use simpler function yaml.safe_load instead of yaml.safe_load_all

Since the configuration .yaml files contain only one document the simpler function
yaml.safe_load can be used. This also allows to get rid of the wrapping for-loop.

slaufmann commented:
Thanks for the additional remarks, I changed the function calls.
